### PR TITLE
remove command and golbal keyword when only root command #1566

### LINKED ISSUE #1566: how to build command without sub command
this is  [official demo with required flag](https://go.dev/play/p/3XapqAxcQOr)

## output
```
NAME:
   play - A new cli application

USAGE:
   play command [command options] [arguments...]

COMMANDS:
   help, h  Shows a list of commands or help for one command

OPTIONS:
   --lang value  language for the greeting (default: "english")
   --help, -h    show help (default: false)
2009/11/10 23:00:00 Required flag "lang" not set

Program exited.
```
## expect
<pre><code>
NAME:
   play - A new cli application

USAGE:
  play  <del>command</del> [<del>command</del> options] [arguments...]

<del>COMMANDS:</del>
 <del>  help, h  Shows a list of commands or help for one command</del>

OPTIONS:
   --lang value  language for the greeting (default: "english")
   --help, -h    show help (default: false)
2009/11/10 23:00:00 Required flag "lang" not set

Program exited.
</code></pre>

### PR DESCRIPTION
**Pull request by y1rn**
## What type of PR is this?

_(REQUIRED)_

- bug

## What this PR does / why we need it:

_(REQUIRED)_

align to the design 

## Which issue(s) this PR fixes:

_(REQUIRED)_

Fix #1566 

## Special notes for your reviewer:

_(fill-in or delete this section)_

## Testing

_(fill-in or delete this section)_

## Release Notes

_(REQUIRED)_

```release-note

```

### CODE REVIEW

File: app_test.go [hunk 1]
```diff
@@ -2060,11 +2057,6 @@ func TestApp_Run_Help(t *testing.T) {
 			hideHelp:      false,
 			wantContains:  "boom - make an explosive entrance",
 		},
-		{
-			helpArguments: []string{"boom", "help"},
```
Comment (dearchap) [inline app_test.go:2064]: we still need to keep this behaviour. If command is not found action should default to helpAction

File: app_test.go [hunk 2]
```diff
@@ -2075,11 +2067,6 @@ func TestApp_Run_Help(t *testing.T) {
 			hideHelp:      true,
 			wantErr:       fmt.Errorf("flag: help requested"),
 		},
-		{
-			helpArguments: []string{"boom", "help"},
```
Comment (dearchap) [inline app_test.go:2079]: We need to keep this

File: godoc-current.txt [hunk 3]
```diff
@@ -35,7 +35,7 @@ var AppHelpTemplate = `NAME:
    {{template "helpNameTemplate" .}}
 
 USAGE:
-   {{if .UsageText}}{{wrap .UsageText 3}}{{else}}{{.HelpName}} {{if .VisibleFlags}}[global options]{{end}}{{if .Commands}} command [command options]{{end}} {{if .ArgsUsage}}{{.ArgsUsage}}{{else}}[arguments...]{{end}}{{end}}{{if .Version}}{{if not .HideVersion}}
```
Comment (dearchap) [inline godoc-current.txt:38]: why not keep global options ?

File: help_test.go [hunk 4]
```diff
@@ -1009,12 +1019,7 @@ func TestHideHelpCommand_False(t *testing.T) {
 		Writer:          ioutil.Discard,
 	}
 
-	err := app.Run([]string{"foo", "help"})
```
Comment (dearchap) [inline help_test.go:1012]: we still need foo help to work

**Comment by dearchap** [comment]
@urfave/cli Please review as this contains some important user facing changes.

File: app_test.go (hunk 1 above)
Comment (y1rn) [reply app_test.go:2064]: I try to remove help command if no sub command.
<pre><code>
	if a.Command(helpCommand.Name) == nil && !a.HideHelp {
		<del>if !a.HideHelpCommand { </del>
		if !a.HideHelpCommand && len(a.Commands) > 0 {
			a.appendCommand(helpCommand)
		}
</code>
</pre>

File: godoc-current.txt (hunk 3 above)
Comment (y1rn) [reply godoc-current.txt:38]: `global` may no need if no sub command

**Comment by remiposo** [comment]
Wouldn't setting `HideHelpCommand` when creating `cli.App` fix the problem?

```go
func main() {
	(&cli.App{
		Name:            "app",
		HideHelpCommand: true,
	}).Run(os.Args)
}
```
```shell
$ ./app
NAME:
   app - A new cli application

USAGE:
   app [global options] [arguments...]

GLOBAL OPTIONS:
   --help, -h  show help (default: false)
```

I think there's room for consideration about the default value of `cli.App.HideHelpCommand` and global keyword when only root command.
In any case, if remove help command, I think it would be natural to implement it by activating HideHelpCommand, not checking length of commands directly!

**Comment by y1rn** [comment]
> Wouldn't setting `HideHelpCommand` when creating `cli.App` fix the problem?
> 
> ```go
> func main() {
> 	(&cli.App{
> 		Name:            "app",
> 		HideHelpCommand: true,
> 	}).Run(os.Args)
> }
> ```
> 
> ```shell
> $ ./app
> NAME:
>    app - A new cli application
> 
> USAGE:
>    app [global options] [arguments...]
> 
> GLOBAL OPTIONS:
>    --help, -h  show help (default: false)
> ```
> 
> I think there's room for consideration about the default value of `cli.App.HideHelpCommand` and global keyword when only root command. In any case, if remove help command, I think it would be natural to implement it by activating HideHelpCommand, not checking length of commands directly!

I think `GLOBAL` is unnecessary if no sub  commands.

**Pull request closed without merging**